
syntax = "proto3";

package descarteslabs.workflows;

import "descarteslabs/common/proto/types/types.proto";
import "descarteslabs/common/proto/typespec/typespec.proto";
import "descarteslabs/common/proto/errors/errors.proto"; // backwards compatible fix for ErrorCode

service XYZAPI {
  rpc CreateXYZ(CreateXYZRequest) returns (XYZ) {}
  rpc GetXYZ(GetXYZRequest) returns (XYZ) {}
  rpc GetXYZSessionErrors(GetXYZSessionErrorsRequest)
      returns (stream XYZError) {}
}

message XYZ {
  string id = 1;
  // milliseconds since the UNIX epoch (UTC)
  int64 created_timestamp = 2;
  // milliseconds since the UNIX epoch (UTC)
  int64 updated_timestamp = 3;
  string name = 7;
  string description = 8;
  ResultType type = 9;
  string channel = 10;
  string client_version = 11;

  string serialized_graft = 21;
  Typespec typespec = 25;

  string user = 23;
  string org = 24;
  // Template string like "https://workflows.descarteslabs.com/v0-5/xyz/1234567/{z}/{x}/{y}.png".
  // Compatible with Leaflet, OpenLayers, Python format strings.
  string url_template = 26;
}

message CreateXYZRequest {
  string name = 1;
  string description = 2;
  string serialized_graft = 3;
  Typespec typespec = 4;
  string channel = 5;
  string client_version = 6;
}

message GetXYZRequest { string xyz_id = 1; }

message GetXYZSessionErrorsRequest {
  string xyz_id = 1;
  string session_id = 2;
  // milliseconds since the UNIX epoch (UTC)
  int64 start_timestamp = 3;
}

message XYZError {
  ErrorCode code = 1;
  string message = 2;
  // milliseconds since the UNIX epoch (UTC)
  int64 timestamp = 3;
  string session_id = 4;
}
